daml-lf/upgrades/src/main/scala/com/digitalasset/daml/lf/upgrades/Typecheck.scala

```diff
@@ -0,0 +1,235 @@
+// Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf
+package upgrades
+
+import com.daml.lf.data.Ref
+import com.daml.lf.language.Ast
+import scala.util.{Try, Success, Failure}
+import com.daml.lf.validation.AlphaEquiv
+
+case class Upgrading[A](past: A, present: A) {
+  def map[B](f: A => B): Upgrading[B] = Upgrading(f(past), f(present))
+  def fold[B](f: (A, A) => B): B = f(past, present)
+  def zip[B, C](that: Upgrading[B], f: (A, B) => C) =
+    Upgrading(f(this.past, that.past), f(this.present, that.present))
+}
+
+case class UpgradeError(message: String) extends Throwable(message)
+
+sealed abstract class UpgradedRecordOrigin
+
+final case class TemplateBody(template: Ref.DottedName) extends UpgradedRecordOrigin
+final case class TemplateChoiceInput(template: Ref.DottedName, choice: Ref.ChoiceName)
+    extends UpgradedRecordOrigin
+final case object TopLevel extends UpgradedRecordOrigin
+
+object Typecheck {
+  def typecheckUpgrades(
+      present: (Ref.PackageId, Ast.Package),
+      mbPast: Option[(Ref.PackageId, Ast.Package)],
+  ): Try[Unit] = {
+    mbPast match {
+      case None => Failure(UpgradeError("CouldNotResolveUpgradedPackageId"));
+      case Some(past) => {
+        val tc = this(Upgrading(past, present))
+        tc.check()
+      }
+    }
+  }
+}
+
+case class Typecheck(packagesAndIds: Upgrading[(Ref.PackageId, Ast.Package)]) {
+  lazy val packageId: Upgrading[Ref.PackageId] = packagesAndIds.map(_._1)
+  lazy val _package: Upgrading[Ast.Package] = packagesAndIds.map(_._2)
+
+  def extractDelExistNew[K, V](
+      arg: Upgrading[Map[K, V]]
+  ): (Map[K, V], Map[K, Upgrading[V]], Map[K, V]) =
+    (
+      arg.past -- arg.present.keySet,
+      arg.past.keySet
+        .intersect(arg.present.keySet)
+        .map(k => k -> Upgrading(arg.past(k), arg.present(k)))
+        .toMap,
+      arg.present -- arg.past.keySet,
+    )
+
+  def checkDeleted[K, V, T <: Throwable](
+      arg: Upgrading[Map[K, V]],
+      handler: V => T,
+  ): Try[(Map[K, Upgrading[V]], Map[K, V])] = {
+    val (deletedV, existingV, newV) = extractDelExistNew(arg)
+    deletedV.headOption match {
+      case Some((k @ _, v)) => Failure(handler(v))
+      case _ => Success((existingV, newV))
+    }
```

Comment:
Currently we take the first missing item on the both the participant side and the compiler side - I'll make a ticket for this, since it's something I'd like to do to both the compiler and assistant.

Reply:
https://github.com/digital-asset/daml/issues/18312